sql-plugin/src/main/scala/com/nvidia/spark/rapids/BatchWithPartitionData.scala

```diff
@@ -0,0 +1,356 @@
+      batch: ColumnarBatch,
+      listOfPartitionedRowsData: Array[Array[PartitionRowData]],
+      partitionSchema: StructType): BatchWithPartitionDataIterator = {
+    withResource(batch) { _ =>
+      val splitIndices = calculateSplitIndices(listOfPartitionedRowsData)
+      closeOnExcept(splitColumnarBatch(batch, splitIndices)) { splitColumnarBatches =>
+        // Combine the split GPU ColumnVectors with partition ColumnVectors.
+        val combinedBatches = splitColumnarBatches.zip(listOfPartitionedRowsData).map {
+          case (spillableBatch, partitionedRowsData) =>
+            new BatchWithPartitionData(spillableBatch, partitionedRowsData, partitionSchema)
+        }
+        new BatchWithPartitionDataIterator(combinedBatches)
+      }
+    }
+  }
+
+  /**
+   * Helper method to calculate split indices for dividing ColumnarBatch based on the row counts.
+   *
+   * Example,
+   * {{{
+   * Input - Array of arrays having (row counts, InternalRow) pairs:
+   * [
+   *  [ (10, [abc, ab]), (40, [bc, ab]), (63, [abc, bc]) ],
+   *  [ (7, [abc, bc]), (10, [aa, cc]), (11, [ade, fd]) ],
+   *  [ (4, [ac, gbc]), (7, [aa, tt]), (15, [ate, afd]) ]
+   * ]
+   *
+   * Result -
+   *  - Row counts for each batch: [ 10+40+30, 7+10+11, 4+7+15 ] = [ 80, 28, 26 ]
+   *  - Split Indices: [ 80, 108 ]
+   *  - ColumnarBatch with 134 rows will be split in three batches having
+   *    `[ 80, 28, 26 ]` rows in each.
+   *}}}
+   *
+   * @param listOfPartitionedRowsData Array of arrays where each entry contains 'list of partition
+   *                                  values and number of times it is replicated'. See example in
+   *                                  `splitPartitionDataIntoGroups()`.
+   * @return A sequence of split indices.
+   */
+  private def calculateSplitIndices(
+      listOfPartitionedRowsData: Array[Array[PartitionRowData]]): Seq[Int] = {
+    // Calculate the row counts for each batch
+    val rowCountsForEachBatch = listOfPartitionedRowsData.map(partitionData =>
+      partitionData.map {
+        case PartitionRowData(_, rowNum) => rowNum
+      }.sum.toInt
+    )
+    // Calculate split indices using cumulative sum
+    rowCountsForEachBatch.scanLeft(0)(_ + _).drop(1).dropRight(1)
+  }
+
+  /**
+   * Split a ColumnarBatch into multiple ColumnarBatches based on indices.
+   */
+  private def splitColumnarBatch(input: ColumnarBatch,
+      splitIndices: Seq[Int]): Seq[SpillableColumnarBatch] = {
+    if (splitIndices.isEmpty) {
+      Seq(SpillableColumnarBatch(GpuColumnVector.incRefCounts(input),
+        SpillPriorities.ACTIVE_ON_DECK_PRIORITY))
+    } else {
+      val schema = GpuColumnVector.extractTypes(input)
+      val splitInput = withResource(GpuColumnVector.from(input)) { table =>
+        table.contiguousSplit(splitIndices: _*)
+      }
+      closeOnExcept(splitInput) { _ =>
+        splitInput.zipWithIndex.safeMap { case (ct, i) =>
+          splitInput(i) = null
```

Comment:
Why is this here?  Should just be a .safeMap without any index to convert each entry to a SpillableColumnarBatch.